greenlight/harness/testcase.py

```diff
@@ -19,10 +19,17 @@ def setUp(self, *args, **kwargs):
         MarionetteTestCase.setUp(self, *args, **kwargs)
         Puppeteer.set_marionette(self, self.marionette)
 
+        self._start_handle_count = len(self.marionette.window_handles)
         self.marionette.set_context('chrome')
         self.browser = self.windows.current
 
     def tearDown(self, *args, **kwargs):
-        MarionetteTestCase.tearDown(self, *args, **kwargs)
+        win_count = len(self.marionette.window_handles)
+        self.assertEqual(win_count, self._start_handle_count,
+                         "A test must not leak window handles. "
+                         "This test started the browser with %s open "
+                         "top level browsing contexts, but ended with %s." %
+                         (self._start_handle_count, win_count))
 
         self.prefs.restore_all_prefs()
+        MarionetteTestCase.tearDown(self, *args, **kwargs)
```

Comment:
I'm still in favor of a `closeAllChromeWindows` method in windows.js, which can be executed in tearDown. It should close all windows beside the last open browser window (optionally added as exception). We do this in our Mozmill tests, and this is pretty failure proof.

Raising an exception as above will not help, giving that we do not close any left open windows. But we play even bad with MarionetteTestCase.tearDown, in case of not executing it.